src/persist/src/indexed/columnar.rs

```diff
@@ -0,0 +1,259 @@
+        debug_assert_eq!(self.timestamps.len(), self.len);
+        debug_assert_eq!(self.diffs.len(), self.len);
+        let key = &self.key_data[self.key_offsets[idx]..self.key_offsets[idx + 1]];
+        let val = &self.val_data[self.val_offsets[idx]..self.val_offsets[idx + 1]];
+        let ts = self.timestamps[idx];
+        let diff = self.diffs[idx];
+        Some(((key, val), ts, diff))
+    }
+
+    /// Iterate through the records in Self.
+    fn iter(&self) -> ColumnarRecordsIter<'a> {
+        ColumnarRecordsIter {
+            idx: 0,
+            records: self.clone(),
+        }
+    }
+}
+
+/// An [Iterator] over the records in a [ColumnarRecords].
+#[derive(Clone, Debug)]
+pub struct ColumnarRecordsIter<'a> {
+    idx: usize,
+    records: ColumnarRecordsRef<'a>,
+}
+
+impl<'a> Iterator for ColumnarRecordsIter<'a> {
+    type Item = ((&'a [u8], &'a [u8]), u64, isize);
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (self.records.len, Some(self.records.len))
+    }
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let ret = self.records.get(self.idx);
+        self.idx += 1;
+        ret
+    }
+}
+
+impl<'a> ExactSizeIterator for ColumnarRecordsIter<'a> {}
+
+/// An abstraction to incrementally add ((Key, Value), Time, Diff) records
+/// in a columnar representation, and eventually get back a [ColumnarRecords].
+#[derive(Debug, Default)]
+pub struct ColumnarRecordsBuilder {
+    records: ColumnarRecords,
+}
+
+impl ColumnarRecordsBuilder {
+    /// The number of (potentially duplicated) ((Key, Val), Time, Diff) records
+    /// stored in Self.
+    pub fn len(&self) -> usize {
+        self.records.len
+    }
+
+    /// Reserve space of `additional` more records, based on `key_size_guess` and
+    /// `val_size_guess`.
+    ///
+    /// The guesses for key and val sizes are best effort, and if they end up being
+    /// too small, the underlying buffers will be resized.
+    pub fn reserve(&mut self, additional: usize, key_size_guess: usize, val_size_guess: usize) {
+        self.records.key_offsets.reserve(additional);
+        self.records.key_data.reserve(additional * key_size_guess);
+        self.records.val_offsets.reserve(additional);
+        self.records.key_data.reserve(additional * val_size_guess);
+        self.records.timestamps.reserve(additional);
+        self.records.diffs.reserve(additional);
+    }
+
+    /// Add a record to Self.
+    pub fn push(&mut self, record: ((&[u8], &[u8]), u64, isize)) {
+        let ((key, val), ts, diff) = record;
+        self.records.key_offsets.push(self.records.key_data.len());
+        self.records.key_data.extend_from_slice(key);
+        self.records.val_offsets.push(self.records.val_data.len());
+        self.records.val_data.extend_from_slice(val);
+        self.records.timestamps.push(ts);
+        self.records.diffs.push(diff);
+        self.records.len += 1;
+        debug_assert_eq!(self.records.key_offsets.len(), self.records.len);
```

Comment:
toss a comment here that this key_offsets.len() and val_offsets.len() checks are different than all the other ones in the file because the final offsets haven't been pushed yet by `finish`